### PR TITLE
feat(autoware_test_utils): add autoware test manager

### PR DESCRIPTION
**Pull request by kminoda**
## Description

I want to add `AutowareTestManager` to minimize the implementation cost of pub-sub based test for general Autoware packages. 

Sample implementation based on this for `detected_object_validation` can be found here: https://github.com/kminoda/autoware.universe/blob/feat/detected_object_validation/add_test/perception/detected_object_validation/test/object_position_filter/test_object_position_filter.cpp

Note that `PlanningInterfaceTestManager` can also be implemented based on this.

## Tests performed

Added a simple test script as well.

## Effects on system behavior

None

## Interface changes

None

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kminoda** [comment]
Code maintainers should be enabled after https://github.com/autowarefoundation/autoware.universe/pull/7456#pullrequestreview-2129497445 is merged to main and this branch.

File: common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp [hunk 1]
```diff
@@ -512,6 +513,62 @@ void publishToTargetNode(
...
+        test_node_, topic_name, callback, subscriber);
+      subscribers_[topic_name] = std::static_pointer_cast<void>(subscriber);
+    } else {
+      RCLCPP_WARN(test_node_->get_logger(), "Subscriber %s already set.", topic_name.c_str());
+    }
+  }
+
+protected:
+  // Publisher
+  std::unordered_map<std::string, std::shared_ptr<void>> publishers_;
```
Comment (soblin) [inline common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp:565]: 1. Please use `rclcpp::PublisherBase` as the base

File: common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp [hunk 2]
```diff
@@ -512,6 +513,62 @@ void publishToTargetNode(
...
+  template <typename MessageType>
+  void test_pub_msg(
+    rclcpp::Node::SharedPtr target_node, const std::string & topic_name, MessageType & msg)
+  {
+    if (publishers_.find(topic_name) == publishers_.end()) {
+      auto publisher = test_node_->create_publisher<MessageType>(topic_name, 10);
+      publishers_[topic_name] = std::static_pointer_cast<void>(publisher);
+    }
+
+    auto publisher =
```
Comment (soblin) [inline common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp:541]: 2. And use `std::dynamic_pointer_cast` to downcast from PublisherBase

**Review by soblin — CHANGES_REQUESTED** [review]

File: common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp (hunk 1 above)
Comment (kminoda) [reply common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp]: Fixed: https://github.com/autowarefoundation/autoware.universe/pull/7597/commits/ed0cee986b7ed8fe3701d68f946358fda19764e7

Comment (kminoda) [reply common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp]: Fixed: https://github.com/autowarefoundation/autoware.universe/pull/7597/commits/4e9c1cadd093f6cdc1333b1d1e707271c08c5b16

File: common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp (hunk 2 above)
Comment (kminoda) [reply common/autoware_test_utils/include/autoware_test_utils/autoware_test_utils.hpp:541]: Fixed: https://github.com/autowarefoundation/autoware.universe/pull/7597/commits/4e9c1cadd093f6cdc1333b1d1e707271c08c5b16

**Review by soblin — APPROVED** [review]

**Pull request merged**